Bug 1815219: CO-876: allow defining Conditions in AWS CredentialsRequest

Danil-Grigorev opened the pull request:
Added support for conditions in AWS provider

Allows to use the same form as JSON conditions to specify rules to apply - https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_condition_operators.html

```yaml
apiVersion: cloudcredential.openshift.io/v1
kind: CredentialsRequest
.....
      - iam:PassRole
      condition:
        "Null":
          key: "true"
      effect: Allow
      resource: '*'
```

Danil-Grigorev commented:
/assign @JoelSpeed

JoelSpeed commented on pkg/apis/cloudcredential/v1/aws_types.go:
```diff
@@ -49,3 +51,71 @@ type AWSProviderStatus struct {
+
+// NotEquals appends NotEquals prefix to the condition
+func (c AWSConditionType) NotEquals() AWSConditionType {
+	return c + "NotEquals"
```
Would a condition type of `StringNotEquals` unmarshal from JSON correctly?

Danil-Grigorev replied:
It totally will, as it will undergo only a raw encoding to internal golang structure, where the types are still raw strings.

JoelSpeed commented on pkg/apis/cloudcredential/v1/aws_types.go:
```diff
@@ -49,3 +51,71 @@ type AWSProviderStatus struct {
+type AWSConditionType string
+
+const (
+	// String elements that restrict access based on comparing a key to a string value.
+	String AWSConditionType = "String"
+	// Numeric elements that restrict access based on comparing a key to an integer or decimal value.
+	Numeric AWSConditionType = "Numeric"
+	// Date elements that restrict access based on comparing a key to a date/time value.
+	Date AWSConditionType = "Date"
+	// Bool elements that restrict access based on comparing a key to "true" or "false."
+	Bool AWSConditionType = "Bool"
+	// Arn elements that restrict access based on comparing a key to an ARN.
+	Arn AWSConditionType = "Arn"
+	// Null check if a condition key is present at the time of authorization.
+	Null AWSConditionType = "Null"
+	// Binary elements that test key values that are in binary format.
+	Binary AWSConditionType = "Binary"
+	// IPAddress restrict access based on comparing a key to an IPv4 or IPv6 address or range of IP addresses.
+	IPAddress AWSConditionType = "IpAddress"
```
Might be worth suffixing these with `ConditionType`? Eg `StringConditionType` to make it more obvious what they are for and also distinguish them from other potentially exported fields that could be added at a later date

Danil-Grigorev replied:
Yeah, makes sense. Will update it, that sounds more cannon.

Danil-Grigorev commented:
/cc @joelddiaz Here is a PR for https://issues.redhat.com/browse/CO-876

abhinavdahiya commented on pkg/apis/cloudcredential/v1/aws_types.go:
```diff
@@ -38,6 +38,8 @@ type StatementEntry struct {
 	// Resource specifies the object(s) this statement should apply to. (or "*" for all)
 	Resource string `json:"resource"`
+	// Condition specifies under which condition StatementEntry will apply
+	Condition AWSCondition `json:"condition,omitempty"`
```
`` Condition AWSCondition `json:"condition,omitempty"` ``  ->  `` Condition StatementEntryCondition `json:"condition,omitempty"` ``

abhinavdahiya commented on pkg/apis/cloudcredential/v1/aws_types.go:
```diff
@@ -49,3 +51,71 @@ type AWSProviderStatus struct {
 }
+
+// AWSCondition - map of condition types, with associated key - value mapping
+type AWSCondition map[AWSConditionType]AWSConditionKeyValue
```
```suggestion
type StatementEntryCondition map[AWSIAMPolicyConditionType]AWSIAMPolicyConditionKeyVakue
```

abhinavdahiya commented on pkg/apis/cloudcredential/v1/aws_types.go:
```diff
@@ -49,3 +51,71 @@ type AWSProviderStatus struct {
+// AWSCondition - map of condition types, with associated key - value mapping
+type AWSCondition map[AWSConditionType]AWSConditionKeyValue
+
+// AWSConditionKeyValue - mapping of values for the chousen type
```
```suggestion
// AWSConditionKeyValue - mapping of values for the chosen type
```

abhinavdahiya commented on pkg/apis/cloudcredential/v1/aws_types.go:
```diff
@@ -49,3 +51,71 @@ type AWSProviderStatus struct {
+type AWSCondition map[AWSConditionType]AWSConditionKeyValue
+
+// AWSConditionKeyValue - mapping of values for the chousen type
+type AWSConditionKeyValue map[string]interface{}
```
```suggestion
type AWSIAMPolicyConditionKeyVakue map[string]interface{}
```

abhinavdahiya commented on pkg/apis/cloudcredential/v1/aws_types.go:
```diff
@@ -49,3 +51,71 @@ type AWSProviderStatus struct {
+type AWSConditionKeyValue map[string]interface{}
+
+// AWSConditionType provides condition operators to match the condition key and value in the policy against values in the request context
+type AWSConditionType string
```
```suggestion
type AWSIAMPolicyConditionType string
```

abhinavdahiya commented on pkg/apis/cloudcredential/v1/aws_types.go:
```diff
@@ -49,3 +51,71 @@ type AWSProviderStatus struct {
+	return "Not" + c
+}
+
+// IgnoreCase appends IgnoreCase prefix to the condition
```
```suggestion
// IgnoreCase appends IgnoreCase suffix to the condition
```

abhinavdahiya commented on pkg/apis/cloudcredential/v1/aws_types.go:
```diff
@@ -49,3 +51,71 @@ type AWSProviderStatus struct {
+	return c + "IgnoreCase"
+}
+
+// Equals appends Equals prefix to the condition
```
```suggestion
// Equals appends Equals suffix to the condition
```

abhinavdahiya commented on pkg/apis/cloudcredential/v1/aws_types.go:
```diff
@@ -49,3 +51,71 @@ type AWSProviderStatus struct {
+	return c + "Equals"
+}
+
+// NotEquals appends NotEquals prefix to the condition
```
```suggestion
// NotEquals appends NotEquals suffix to the condition
```

abhinavdahiya commented on pkg/apis/cloudcredential/v1/aws_types.go:
```diff
@@ -49,3 +51,71 @@ type AWSProviderStatus struct {
+	return c + "NotEquals"
+}
+
+// Like appends Like prefix to the condition
```
```suggestion
// Like appends Like suffix to the condition
```

abhinavdahiya commented on pkg/apis/cloudcredential/v1/aws_types.go:
```diff
@@ -49,3 +51,71 @@ type AWSProviderStatus struct {
+	return c + "Like"
+}
+
+// NotLike appends NotLike prefix to the condition
```
```suggestion
// NotLike appends NotLike suffix to the condition
```

abhinavdahiya commented on pkg/apis/cloudcredential/v1/aws_types.go:
```diff
@@ -49,3 +51,71 @@ type AWSProviderStatus struct {
+	return c + "NotLike"
+}
+
+// LessThan appends LessThan prefix to the condition
```
```suggestion
// LessThan appends LessThan suffix to the condition
```

abhinavdahiya commented on pkg/apis/cloudcredential/v1/aws_types.go:
```diff
@@ -49,3 +51,71 @@ type AWSProviderStatus struct {
+	return c + "LessThan"
+}
+
+// GreaterThan appends GreaterThan prefix to the condition
```
```suggestion
// GreaterThan appends GreaterThan suffix to the condition
```

Danil-Grigorev commented:
@joelddiaz Sorry, looking at the PR now, I can't see a reason for presence of AWS IAM condition types, or it will actually be a total overhead for the scope of the PR. Agree with you, going to remove unnecessary parts and change naming.

Danil-Grigorev commented:
@joelddiaz Tested this PR in my setup, works fine for any combination of conditions. I addressed concerns around constants, and decided to remove them, as it sounded clear that is not the thing needed for a static template provided externally. @JoelSpeed I hope this integration is enough to cover the needs for the conditions. In case the tests are still needed to display the usage, I have a sample part to update the PR on request, but aside from that the tests are of no use here. This PR is sufficient in it's stage, as it is only backfilling missing fields of supported template.

enxebre commented:
I'm retitling to make sure this is tracked along with the BZ issue. Same need to be done with the upcoming counterpart PR for the machine API Operator.
/retitle Bug 1815219: CO-876: allow defining Conditions in AWS CredentialsRequest

JoelSpeed commented on pkg/apis/cloudcredential/v1/aws_types.go:
```diff
@@ -49,3 +51,9 @@ type AWSProviderStatus struct {
 	Policy string `json:"policy"`
 }
+
+// AWSIAMCondition - map of condition types, with associated key - value mapping
+type IAMCondition map[string]IAMPolicyConditionKeyValue
```
We want this to be consistent with other types, so either `IAMPolicyCondition` or `StatementEntryCondition` I would have said
```suggestion
// IAMPolicyCondition - map of condition types, with associated key - value mapping
type IAMPolicyCondition map[string]IAMPolicyConditionKeyValue
```

JoelSpeed commented:
I've manually verified that this works with https://github.com/openshift/machine-api-operator/pull/557 to create the expected policy on the user

enxebre commented:
lgtm as well @joelddiaz you folks good to merge?

enxebre commented:
thanks @Danil-Grigorev ping @joelddiaz :)

dgoodwin commented on pkg/apis/cloudcredential/v1/aws_types.go:
```diff
@@ -38,6 +38,8 @@ type StatementEntry struct {
 	// Resource specifies the object(s) this statement should apply to. (or "*" for all)
 	Resource string `json:"resource"`
+	// Condition specifies under which condition StatementEntry will apply
+	Condition IAMPolicyCondition `json:"condition,omitempty"`
```
I'd prefer "iamPolicyCondition" or "policyCondition" to "condition" for the json key. Condition collides a little too much with typical status.conditions.

Danil-Grigorev replied:
Ok, fixing now.

Danil-Grigorev commented on pkg/aws/actuator/actuator.go:
```diff
@@ -1154,9 +1155,10 @@ type PolicyDocument struct {
+	Effect    string
+	Action    []string
+	Resource  string
+	Condition minterv1.IAMPolicyCondition `json:",omitempty"`
```
No, it is necessary, as an empty condition in the template is not accepted by AWS validation

dgoodwin commented:
/lgtm
